### PR TITLE
ISSUE-#154 High priority, supporting memory for preserving favorited courseIds

### PR DESCRIPTION
**Pull request by heatsink**
The courses you favorite will stay preserved even when closing the app.

Additionally, you will now start with 0 courses favorited upon first boot-up.

### CODE REVIEW

**Comment by heatsink** [comment]
Hang on for merge, I'll add some tests.

**Comment by heatsink** [comment]
There we go, now we only decrease to 74% code coverage instead of >69% code coverage lol

**Comment by AndiAndreas** [comment]
There is a Problem in your code. When you start up the app there arent any courses. I then went into debugger and found out that you get mock data on first start up of timeTable. When you oben TimeTable again there is the prod data. This even happens if you wait some time on login page after starting the app.

**Comment by AndiAndreas** [comment]
When you log in it works. But for guest it does not work.

**Comment by heatsink** [comment]
Thanks for noticing this. I made a fix, but I'll make it a bit cleaner and post it here tomorrow.

**Comment by heatsink** [comment]
On start up, there is a slight flicker for refresh regarding the timetable as the widgets propagate their build methods. I'll make a low priority issue for it.

File: lib/utils/schedulingUtility.dart [hunk 1]
```diff
@@ -31,12 +41,22 @@ class SchedulingUtility {
...
 
   static String constructSchedulingConflictText(Lecture one, Lecture two) {
+    int startOne = convertToMinutes(one.startDayTime.hour,
+        one.startDayTime.minute);
+    int endOne = convertToMinutes(one.endDayTime.hour,
+        one.endDayTime.minute);
+    int startTwo = convertToMinutes(two.startDayTime.hour,
+        two.startDayTime.minute);
+    if (startTwo >= startOne && startTwo <= endOne) {
+      return "Please consider that the schedules between " + one.course.name +
```
Comment (RedTo) [inline lib/utils/schedulingUtility.dart]: Good one, I just wanted to open an issue for this... Great if you already solved it.

**Comment by RedTo** [comment]
The code looks good, but I found a logical bug.

Assuming, that a person has added a course to her/his favorites and afterwards the course would be deleted out of the courselist, it will still be present in the favorites, right?

So I'm not quite sure if this is what we should do. More or less I would say, that we should only show courses that are available in the courselist. Otherwise we have to thing about to show a notification or something similar if the course is gone from courselist...

I'm not sure how and if we should handle all this stuff, or just take your solution?!
Please respond: @heatsink @AndiAndreas

**Comment by heatsink** [comment]
True, I haven't implemented anything to worry about updates to Nine.

You are correct that in the current implementation, someone who favorites a course will still have that course favorited if courses are removed from the Nine API.

The proper fix would be to implement versioning (and optionally ask the professor working with Nine to return a link like https://nine.wi.hm.edu/api/CIE/Versions which just has a list of versions such as [{v1: (dateReleased)}, {v2: (dateReleased)}, {v3: (dateReleased}). Then, we could do support more advanced logic because I think it would be confusing to have a course favorited but have it disappear completely the next day if the API gets updated.

If the professor doesn't want to do that, then we could ourselves maintain a list of API links and we could do the same advanced logic to show courses as expired, out of date, or canceled if it is in one version but not the next for the same semester.

I don't know how the professor running Nine does things, but I would expect that the data in https://nine.wi.hm.edu/api/v2/courses/FK%2013/CIE/SoSe%2018 would never change and that new changes would be present in a new API resource such as https://nine.wi.hm.edu/api/**v3**/courses/FK%2013/CIE/SoSe%2018.

**Comment by heatsink** [comment]
However, in the interest of time, I think we should leave the proper fix I mentioned out of this sprint. Since I'm not sure how the professor would handle updates to the Nine API, I think we shouldn't worry about it yet. 

The reason being, if the Nine API is updated by changing the links with a new version (ex. v2 to v3), then we wouldn't have to worry about favorited courses being out of date since we would have a new, bigger problem: We wouldn't be using the new link and all of our courses would be out of date.

If anyone knows how the API might be updated, please let me know! If the contents of https://nine.wi.hm.edu/api/v2/courses/FK%2013/CIE/SoSe%2018 itself would be modified, then I think I will quickly implement the temporary solution suggested by @RedTo.

**Comment by RedTo** [comment]
Hmm. I just thought about it again and ask me, if the courses will change
after they are in the nine DB. So maybe we just go with your solution. A
fix might also be to check if they are in course list and Mark them gray or
something like this. They should also not be able to submit later...

Sean Wallace <notifications@github.com> schrieb am Mi., 20. Juni 2018,
18:03:

**Comment by RedTo** [comment]
@AndiAndreas  what do you think? Should we just use this approach or add some additional functionality?

Oh, I got another idea: take this solution and add a issue with optional and close it. All those issues we can implement if Ms Socher will choose our app. (Don't forget to close them, so we have not open that many at the end)

**Comment by RedTo** [comment]
I hope it is okay, that I have fixed some problems on this branch (mainly merging with master)...

Next step would be to pull this stuff over to master and after that I will then make the ics export running.

**Pull request merged**